### ISSUE TITLE
RM Blog post on Reactive Messaging

### ISSUE THREAD

**Issue by Emily-Jiang** [OpenLiberty/open-liberty#6682]
Write bog posts as documented by WAD

**Comment by Azquelt** [comment]
Potential blog content (largely adapted from Gordon's beta blog post section):

An application using reactive messaging is composed of CDI beans consuming, producing, and processing messages passing along reactive streams. These messages can be internal to the application or can be sent and received via different message brokers.

Reactive Messaging provides a very easy to use way to send, receive, and process messages. With MicroProfile Reactive Messaging, you can annotate application beans' methods to have messages on a particular channel (`@Incoming`, `@Outgoing`, or both) and Liberty drives those methods appropriately as reactive streams publishers, subscribers, or processors.

To enable the feature include it in your `server.xml` feature list:

[source,xml]
----
<featureManager>
  <feature>mpReactiveMessaging-1.0</feature>
  ...
</featureManager>
----

With this feature in the OpenLiberty runtime, an application CDI bean can have one of its methods annotated as being message driven. In the example below, the method processes messages from the "greetings" channel:

[source,java]
----
@Incoming("greetings")
public CompletionStage<Void> consume(Message<String> greeting) {
   return greeting.ack();
}
----

== Using connectors

A channel represents a stream of messages of a given type and, usually, the same topic. Channels can operate locally within the process or use _connectors_ to send messages between services.

For example, with no code changes we could use the `io.openliberty.kafka` connector included in OpenLiberty to change the consume method above to subscribe to messages from the Kafka greetings topic by adding the following config:

[source,text]
----
mp.messaging.incoming.greetings.connector=io.openliberty.kafka
mp.messaging.incoming.greetings.bootstrap.servers=kafkabrokerhost:9092
----

These properties can be set anywhere that's read by MicroProfile Config, for example as system properties via OpenLiberty's `bootstrap.properties` file or environment variables from OpenLiberty's `server.env` file.

The Reactive Messaging specification defines two ways to set configuration properties for a connector:

[source,text]
----
# To set a property only for the connector on this channel
mp.messaging.incoming.greetings.[PROPERTY-NAME]=value1

# To set a property for every channel which uses the io.openliberty.kafka connector
mp.messaging.connector.io.openliberty.kafka.[PROPERTY-NAME]=value2
----

Properties supported by the `io.openliberty.kafka` connector for incoming channels
[options="header",cols="1,2"]
|===
|Property   |Description   
//-------------
|`topic`   |The name of the topic to subscribe to
|`unacked.limit` | The number outstanding unacknowledged messages. If this limit is reached, the connector will stop retrieving records from Kafka until some messages have been acknowledged.
|<any other property>   |All other properties are passed directly as config parameters to the `KafkaConsumer` API. A list of valid properties can be found http://kafka.apache.org/documentation.html#consumerconfigs[in the kafka documentation].
|===

Properties supported by the `io.openliberty.kafka` connector for outgoing channels
[options="header",cols="1,2"]
|===
|Property   |Description   
//-------------
|`topic`   |The name of the topic to send messages to
|<any other property>   |All other properties are passed directly as config parameters to the `KafkaProducer` API. A list of valid properties can be found http://kafka.apache.org/documentation.html#producerconfigs[in the kafka documentation].
|===

So, for example, a full set of properties to connect several channels to topics in IBM Public Cloud Event Streams might look like this:

[source,text]
----
mp.messaging.incoming.greetings.connector=io.openliberty.kafka
mp.messaging.outgoing.new-updates.connector=io.openliberty.kafka

mp.messaging.connector.io.openliberty.kafka.bootstrap.servers=broker-1-eventstreams.cloud.ibm.com:9093,broker-2-eventstreams.cloud.ibm.com:9093
mp.messaging.connector.io.openliberty.kafka.sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="token" password="my-apikey";
mp.messaging.connector.io.openliberty.kafka.sasl.mechanism=PLAIN
mp.messaging.connector.io.openliberty.kafka.security.protocol=SASL_SSL
mp.messaging.connector.io.openliberty.kafka.ssl.protocol=TLSv1.2
----

== Application packaging

When using Kafka-based channels, Open Liberty Reactive Messaging 1.0 loads the Kafka client classes using the application classloader. If you are using the `io.openliberty.kafka` connector to
read or write Kafka messages, include in your application a Kafka client API jar that is compatible with your Kafka server. For example, the `/WEB-INF/lib/` folder would be a suitable place to place a Kafka client JAR when building the application's `.war` file.

Find out more in the link:https://download.eclipse.org/microprofile/microprofile-reactive-messaging-1.0/microprofile-reactive-messaging-spec.pdf[MicroProfile Reactive Messaging spec].

**Comment by Azquelt** [comment]
Note, we need to check that this content is still correct once #8551 is complete.

**Comment by tevans78** [comment]
1. need to change connector id from `io.openliberty.kafka` to `liberty-kafka`
2. mention that topic defaults to channel name
3. perhaps should use microprofile-config.properties file as an example source rather than server.env
4. should put something in about the LibertyLoginModule

**Comment by tevans78** [comment]
5. Add a section on logging and trace. Mention pulling in the slf4j jdk14 jar along with the kafka client and then setting JUL trace as normal for liberty, e.g. in boostrap.properties
```
com.ibm.ws.logging.trace.specification = *=info:logservice=detail:REACTIVEMESSAGE=all:org.apache.kafka.*=WARNING
```

**Comment by gcharters** [comment]
Would be nice to have an `@Outgoing` example.

**Comment by Azquelt** [comment]
We could do an outgoing example, though it's non-trivial. The example you've given will spam out "Message in a bottle" messages as fast as they're consumed forever.

**Comment by Azquelt** [comment]
We should definitely have an example with both `@Incoming` and `@Outgoing` (receiving messages, doing some processing, outputting a new message).

**Comment by Azquelt** [comment]
I've made some substantial updates to the gist linked above to include the content requested by @gcharters and @tevans78 .

Two points that aren't currently mentioned are
* Message acknowledgement
* Relationship to Reactive Streams and MP Reactive Operators

I'm unsure if it's worth going into detail on these topics or not.

**Comment by Emily-Jiang** [comment]
@Azquelt please mention the above two items in the first blog post, as I think they are important.

**Comment by Azquelt** [comment]
@Emily-Jiang I've updated the gist to include message acknowledgement and the relationship to other reactive specs.

**Comment by lauracowen** [comment]
@Azquelt Could you move some of the info from the end section (about the specs) to the introduction so that the reader gets context about reactive streams and why they might be moving messages back and forth up-front? Can link to the specs from there and then don't need that last section. Otherwise, looks good. (And create a draft blog post in https://github.com/OpenLiberty/blogs). Thanks

**Comment by tevans78** [comment]
Blog post is now live.
